src/math.F90

```diff
@@ -659,4 +675,90 @@ function watt_spectrum(a, b) result(E_out)
 
   end function watt_spectrum
 
+!===============================================================================
+! W acts as a front end to the MIT Faddeeva function, Faddeeva_w.
+!===============================================================================
+
+  subroutine w(Z,wv)
+    use ISO_C_BINDING
+    complex(C_DOUBLE_COMPLEX), intent(inout) :: Z ! The point to evaluate Z at
+    complex(8), intent(out) :: wv                 ! The resulting w(Z) value
+    real(C_DOUBLE) :: relerr ! Target relative error in inner loop of MIT Faddeeva
+    logical        :: mangle ! Do we need to perform the special mangling step?
+
+    ! There is some special mangling the true W function in WHOPPER does.  I
+    ! replicate it here.
+    if(aimag(Z) < 0) then
+      Z = conjg(Z)
+      mangle = .TRUE.
+    else
+      mangle = .FALSE.
+    end if
+
+    ! Calculate Faddeeva
+    relerr = 0.0_8
+    wv = faddeeva_w(Z,relerr)
+
+    ! Finish mangling the results
+    if(mangle .eqv. .TRUE.) then
+      wv = -conjg(wv)
+    end if
+  end subroutine
+
+!===============================================================================
+! BROADEN_N_POLYNOMIALS doppler broadens polynomials of the form
+! a/En + b/sqrt(En) + c + d sqrt(En) ...
+! exactly and quickly.
+!===============================================================================
+
+  subroutine broaden_n_polynomials(En, DOPP, n, factors)
+    real(8), intent(in) :: En         ! Energy to evaluate at
+    real(8), intent(in) :: DOPP       ! sqrt(atomic weight ratio / kT), kT given in eV.
+    integer, intent(in) :: n          ! number of components to polynomial
+    real(8), intent(out):: factors(n) ! output leading coefficient
+
+    integer :: i
+
+    real(8) :: sqrtE
+    real(8) :: beta
+    real(8) :: halfinvDOPP2
+    real(8) :: quarterinvDOPP4
+    real(8) :: erfbeta
+    real(8) :: exp_m_beta2
+
+    sqrtE = sqrt(En)
+    beta = sqrtE*DOPP
+    halfinvDOPP2 = 0.5_8/DOPP**2
+    quarterinvDOPP4 = 0.25_8/DOPP**4
+
+    if (beta > 6.0_8) then
+      ! Save time, ERF(6) is 1 to machine precision.
+      ! beta/sqrtpi*exp(-beta**2) is also approximately 1 machine epsilon.
+      erfBeta = 1.0_8
+      exp_m_beta2 = 0.0_8
+    else
+      erfBeta = erf(beta)
+      exp_m_beta2 = exp(-beta**2)
+    end if
+
+    ! Assume that, for sure, we'll use a second order (1/E, 1/V, const)
+    ! fit, and no less.
+
+    factors(1) = erfbeta/En
+    factors(2) = 1.0_8/sqrtE
```

Comment:
There is a risk when using literal constants that you may forget the `_8` to get a double, which may result in lost precision, so getting in the habit of using these constants mitigates that risk somewhat.